// luna-auth/bootstrap/admin-actions.js
'use strict'

module.exports = {
  actions: [
    {
      uid: 'webhooks.create',
      displayName: 'Create',
      pluginName: 'auth',
      section: 'settings',
      category: 'webhooks'
    },
    {
      uid: 'webhooks.read',
      displayName: 'Read',
      pluginName: 'auth',
      section: 'settings',
      category: 'webhooks'
    },
    {
      uid: 'webhooks.update',
      displayName: 'Update',
      pluginName: 'auth',
      section: 'settings',
      category: 'webhooks'
    },
    {
      uid: 'webhooks.delete',
      displayName: 'Delete',
      pluginName: 'auth',
      section: 'settings',
      category: 'webhooks'
    },
    {
      uid: 'users.create',
      displayName: 'Create (invite)',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'users'
    },
    {
      uid: 'users.read',
      displayName: 'Read',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'users'
    },
    {
      uid: 'users.update',
      displayName: 'Update',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'users'
    },
    {
      uid: 'users.delete',
      displayName: 'Delete',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'users'
    },
    {
      uid: 'roles.create',
      displayName: 'Create',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'roles'
    },
    {
      uid: 'roles.read',
      displayName: 'Read',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'roles'
    },
    {
      uid: 'roles.update',
      displayName: 'Update',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'roles'
    },
    {
      uid: 'roles.delete',
      displayName: 'Delete',
      pluginName: 'auth',
      section: 'settings',
      category: 'users and roles',
      subCategory: 'roles'
    }
  ]
}
